### PR TITLE
Allow customization of libraries by server operators

### PR DESCRIPTION
**Pull request by smartboyathome**
This allows server operators to manually specify libraries that their plugins need to run. Phase one of #451.

### CODE REVIEW

File: src/main/java/net/glowstone/util/Library.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
+package net.glowstone.util;
```
Comment (mastercoms) [inline src/main/java/net/glowstone/util/Library.java]: Perhaps move these into a new package?

File: src/main/java/net/glowstone/GlowServer.java [hunk 2]
```diff
@@ -122,8 +122,8 @@
 import net.glowstone.util.GlowHelpMap;
 import net.glowstone.util.GlowServerIcon;
 import net.glowstone.util.GlowUnsafeValues;
-import net.glowstone.util.Library;
-import net.glowstone.util.LibraryManager;
+import net.glowstone.library.Library;
+import net.glowstone.library.LibraryManager;
```
Comment (mastercoms) [inline src/main/java/net/glowstone/GlowServer.java]: Sorry that I wasn't clear, I meant net.glowstone.util.library, but I'm fine with this too. It's really up to you with which you feel is better.

File: src/main/java/net/glowstone/util/ClassPathAgent.java [hunk 3]
```diff
@@ -11,7 +11,7 @@ public static void agentmain(String agentArgs, Instrumentation instrumentation)
         inst = instrumentation;
     }
 
-    static void addJarFile(JarFile file) {
+    public static void addJarFile(JarFile file) {
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/ClassPathAgent.java]: Needs documentation

File: src/main/java/net/glowstone/util/config/ServerConfig.java [hunk 4]
```diff
@@ -134,6 +138,24 @@ public boolean getBoolean(Key key) {
         return bool;
     }
 
+    /**
+     * Retrieves a section as a list of maps.
+     * @param key the key to look up
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/config/ServerConfig.java]: Blank line before description and params

File: src/main/java/net/glowstone/util/library/Library.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    private static final String ARTIFACT_ID_KEY = "artifact-id";
+    private static final String CHECKSUM_KEY = "checksum";
+    private static final String CHECKSUM_TYPE_KEY = "type";
+    private static final String CHECKSUM_VALUE_KEY = "value";
+    private static final String GROUP_ID_KEY = "group-id";
+    private static final String REPOSITORY_KEY = "repository";
+    private static final String VERSION_KEY = "version";
+
+    @SuppressWarnings("unchecked")
+    public static Library fromConfigMap(Map<?, ?> configMap) {
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/library/Library.java]: Needs documentation

File: src/main/java/net/glowstone/util/library/Library.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                   String repository, HashAlgorithm checksumType, String checksumValue) {
+        this.groupId = groupId;
+        this.artifactId = artifactId;
+        this.version = version;
+        this.repository = repository;
+        this.checksumType = checksumType;
+        this.checksumValue = checksumValue;
+    }
+
+    public String getGroupId() {
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/library/Library.java]: Lombokify these getters

File: src/main/java/net/glowstone/util/library/Library.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    public HashAlgorithm getChecksumType() {
+        return checksumType;
+    }
+
+    public String getChecksumValue() {
+        return checksumValue;
+    }
+
+    public Map<?, ?> toConfigMap() {
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/library/Library.java]: Needs documentation

**Review by aramperes — CHANGES_REQUESTED** [review]

File: src/main/java/net/glowstone/util/ClassPathAgent.java [hunk 8]
```diff
@@ -11,7 +11,11 @@ public static void agentmain(String agentArgs, Instrumentation instrumentation)
         inst = instrumentation;
     }
 
-    static void addJarFile(JarFile file) {
+    /**
+     * Adds a JAR file to the system class loader.
+     * @param file The JAR file to add to the class loader.
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/ClassPathAgent.java]: Line between the description and params

**Review by aramperes — CHANGES_REQUESTED** [review]

File: src/main/java/net/glowstone/util/library/LibraryManager.java [hunk 9]
```diff
@@ -227,15 +221,16 @@ String getLibrary() {
          * the checksum will be automatically validated.
          *
          * @param file the file.
-         * @param checksum the reference checksum to validate the file's digest with.
-         * @param algorithm the {@link HashAlgorithm} used to compute the file's digest.
+         * @param library the library containing the algorithm and the checksum.
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/library/LibraryManager.java]: `the {@link Library} containing [...]`

File: src/main/java/net/glowstone/util/library/LibraryManager.java [hunk 10]
```diff
@@ -102,30 +97,25 @@ public void run() {
...
          * Creates an instance of the downloader for a library.
          *
-         * @param group the Maven group ID of the library
-         * @param library the Maven artifact ID of the library
-         * @param version the Maven artifact version of the library
-         * @param checksum the reference checksum of the library.
-         *                  See {@link LibraryDownloader#checksum(File, String, HashAlgorithm)}
-         * @param algorithm the algorithm used to validate the checksum.
-         *                   See {@link LibraryDownloader#checksum(File, String, HashAlgorithm)}
+         * @param library an object representing a library
```
Comment (aramperes) [inline src/main/java/net/glowstone/util/library/LibraryManager.java]: `a {@link Library} instance representing a library`

**Review by aramperes — CHANGES_REQUESTED** [review]

**Review by aramperes — APPROVED** [review]

**Review by mastercoms — APPROVED** [review]

**Pull request merged**